### DISCUSSION TITLE
How can the PR author avoid their PR stalling?

### DISCUSSION THREAD

**Discussion by EvidentlyCube** [TiddlyWiki/TiddlyWiki5#7044] [General]
Related to thread [#6572: For potential users: Are the number of "open issues" and "open pull requests" an issue?](https://github.com/Jermolene/TiddlyWiki5/discussions/6572)

At the moment I have 4 open PRs, one of them has not had any input since it was created 26 days ago. I understand the time of the people who do the reviewing and so I want to do my best to make the process simpler and more pleasant for the reviewers. 

I am going to guess these actions improve the process:

* Implement one change per PR to make it easier to review and understand.
* Add tests wherever possible.
* Be descriptive in the PR description.

But is there anything else? Any formal/informal rules when it makes most sense to ping someone and whom to ping if a PR has stalled? What's the expected amount of time one should wait before actually worrying the PR has been forgotten?

**Comment by saqimtiaz** [comment]
In general it is worth keeping in mind that this is a small project and all PRs ultimately do need to be reviewed by Jeremy, which slows things down but also contributes to the stability and backwards compatibility of TiddlyWiki. Having been involved with TW on and off for almost 15 years I would say that as of the last 3 years in particular Jeremy has been on top of things far more often than not.

 I would probably say that how long to wait depends a bit and what follows is entirely my own opinion. Under normal circumstances maybe two to three weeks before pinging Jeremy for feedback, that is outside of holiday season (winter and summer). Another time that PRs don't get reviewed quickly or merged is when we are close to a TW release as we are now. All the but most trivial PRs should not be merged very close to the release process to ensure sufficient time for user testing.

You can also ping other developers such as @pmario and myself to get feedback and insight from those more familiar with the core until Jeremy has a chance to review.

> * Implement one change per PR to make it easier to review and understand.
> * Add tests wherever possible.
> * Be descriptive in the PR description.

All of these are good ideas and definitely help. For all the but the most trivial PRs it can also be very helpful to first create an issue to propose a change and discuss it before putting in further time and effort, to make sure that the work being done is aligned with the vision for the core.

**Reply by EvidentlyCube** [reply]
>  Another time that PRs don't get reviewed quickly or merged is when we are close to a TW release as we are now. 

Oh no, I need to make PL language fixes asap! Also a suggestion - when a new release is being prepared a pinned Discussion thread could be created to make it obvious it's happening (and for any last minute call to arms).

>  [...] it can also be very helpful to first create an issue to propose a change and discuss it [...]

That's true, it's just that I am a man of action. Or rather once I hyperfixate on something it's faster and simpler for me to do the coding and get it out of my head rather than prolong the process. And I often find it easier to explain things with an actual example. But I'll keep that in mind.

**Reply by saqimtiaz** [reply]
> Oh no, I need to make PL language fixes asap! Also a suggestion - when a new release is being prepared a pinned Discussion thread could be created to make it obvious it's happening (and for any last minute call to arms).

Translation changes should not face any problems in getting merged late, code changes are tricky unless they are very basic. We almost always create a discussion thread and an issue impending an upcoming release, it has just slipped through the cracks and has not happened this time. This is an exception though and not the rule.

**Reply by EvidentlyCube** [reply]
> Translation changes should not face any problems in getting merged late

Still, I want to get those changes in before the release actually happens, otherwise they'll have to wait a while :).

**Comment by pmario** [comment]
I think you are on a good way. I've looked at your PRs and they have everything which makes it possible to be merged. ... One of the biggest strength of a contributor for TW is "patience".

As far as I know, Jeremy uses his e-mail client to track issues, PRs, Talk-TiddlyWiki threads and may be other sources. So you probably can imagine how the inbox may look like. It can happen, that things gets delayed because of the amount of mails that have to be read. 

As Saq wrote you can also ping him or me to have a review if we have time. ... 

But to be honest, I have to say there is also "personal interest". If I use something that a pending PR tries to fix, I'm motivated to download the branch and test the stuff. If a PR contains a fix for something that I personally don't use or that I don't like it's really painful to invest the time. ... But I do love a good story, why something is important ;)

**Reply by EvidentlyCube** [reply]
> One of the biggest strength of a contributor for TW is "patience".

I figured it will be my mortal enemy!

> [...]"personal interest" [...] I do love a good story [...]

Good points, so:
* Prepare a demo of the PR in action.
* Clearly and explain **What** is the PR fixing and **Why** you want it fixed

**Reply by saqimtiaz** [reply]
> * Prepare a demo of the PR in action.

Especially for PRs that impact UI/UX this is incredibly helpful. More often that not, I review PRs and provide feedback between other tasks that take up my time, so a PR that I have to checkout and build will take far longer for me to get to.

**Comment by Jermolene** [comment]
Hi @EvidentlyCube first, apologies for the delays. I can appreciate that it can be frustrating. I humbly appreciate the privileged position I am in where others wish to contribute to the project, and I want to treat contributors accordingly.

Obviously in some sense the primary cause of the problems is that I don't spend enough time on PRs on a regular basis. But conversely, a paradoxical challenge is that reviewing and processing PRs is significantly more time consuming and demanding than writing code myself. I've plenty of experience of merging PRs out of feelings of guilt, and then regretting it.

So, I've learned to try to relax about the backlog. As your [figures show](https://github.com/Jermolene/TiddlyWiki5/discussions/6572#discussioncomment-4176164), the majority of PRs do get merged in a timely fashion. It's more important for me to avoid mistakes than to increase velocity.

As to what people can do, the first thing is not to be afraid of bumping a neglected PR, especially if I haven't commented on it.

It would also help if PRs observed the rules in [contributing.md](https://github.com/Jermolene/TiddlyWiki5/blob/master/contributing.md). So, so many times a poorly written description leaves me reading through the code to try to figure out what the PR is intended to do. Or we often get visual changes without an accompanying before and after screenshot.

**Comment by linonetwo** [comment]
Maybe an auto deployed preview and a link to deployed playground will speed up the review and make review more fun?

An example is https://github.com/tiddly-gittly/TiddlyWiki-CPL/pull/81
![截屏2022-11-24 22 46 33](https://user-images.githubusercontent.com/3746270/203811800-cdc3ab2f-f5b1-4bf3-8142-456d70ed54e5.png)

(It is 404 now because PR has been merged.)

Seems in .github/workflows/ci.yml we already build HTML file for each PR? Then you can config vercel to create preview deploy for this HTML for free. Just need owner of the repo (Jeremy) do some setup in the vercel.

**Comment by Jermolene** [comment]
> Maybe an auto deployed preview and a link to deployed playground will speed up the review and make review more fun?

I'd be delighted to have something like that set up. Would our usage fall into Vercel's free tier?

> Seems in .github/workflows/ci.yml we already build HTML file for each PR? Then you can config vercel to create preview deploy for this HTML for free. Just need owner of the repo (Jeremy) do some setup in the vercel.

We only build the full site for the "master" and "tiddlywiki-com" branches. The reason that we don't build every branch is that we deploy to a single GitHub Pages site which would get unwieldy. There are also potential security issues with building every PR and deploying to the tiddlywiki.com domain. So I'd rather that the branch previews were on another domain entirely.

**Reply by linonetwo** [reply]
This deploy happened on vercel side, and domain is provided by vercel, we can also choose netlify or other service, they will work in same way.

I think opensource project can use them for free, like my repos under tiddly-gittly github organization. You can try config it once you have time, maybe only you have the permission.

**Reply by Jermolene** [reply]
Thanks @linonetwo I've set things up with Vercel:

https://vercel.com/jermolene/tiddlywiki5/deployments
